protocol/message_persistence.go

```diff
@@ -2207,3 +2181,161 @@ func (db sqlitePersistence) SetContactRequestState(id string, state common.Conta
+		// common.Message
+		var message common.Message
+		if err := db.tableUserMessagesScanAllFields(rows, &message); err != nil {
+			return nil, err
+		}
+
+		if msg, ok := messageIdx[message.ID]; !ok {
+			messageIdx[message.ID] = &message
+		} else if discordMessage := msg.GetDiscordMessage(); discordMessage != nil {
+			msg.Payload = getUpdatedChatMessagePayload(discordMessage, message.GetDiscordMessage())
+			messageIdx[message.ID] = msg
+		}
+	}
+
+	var messages []*common.Message
+	for _, message := range messageIdx {
+		messages = append(messages, message)
+	}
+	sort.Sort(sort.Reverse(ByClock{messages}))
+
+	return messages, nil
+}
+
+func getMessagesAndCursorsFromScanRows(db sqlitePersistence, rows *sql.Rows) ([]*common.Message, []string, error) {
+
+	var cursors []string
+	messageIdx := make(map[string]*common.Message, 0)
+
+	for rows.Next() {
+		// There's a possibility of multiple rows per message if the
+		// message has a discordMessage and the discordMessage has multiple
+		// attachments
+		//
+		// Hence, we make sure we're aggregating all attachments on a single
+		// common.Message
+		var (
+			message common.Message
+			cursor  string
+		)
+		if err := db.tableUserMessagesScanAllFields(rows, &message, &cursor); err != nil {
+			return nil, nil, err
+		}
+
+		if msg, ok := messageIdx[message.ID]; !ok {
+			messageIdx[message.ID] = &message
+			cursors = append(cursors, cursor)
+		} else if discordMessage := msg.GetDiscordMessage(); discordMessage != nil {
+			msg.Payload = getUpdatedChatMessagePayload(discordMessage, message.GetDiscordMessage())
+			messageIdx[message.ID] = msg
+		}
+	}
+
+	var messages []*common.Message
+	for _, message := range messageIdx {
+		messages = append(messages, message)
+	}
+	sort.Sort(sort.Reverse(ByClock{messages}))
+
+	return messages, cursors, nil
+}
+
+func getPinnedMessagesAndCursorsFromScanRows(db sqlitePersistence, rows *sql.Rows) ([]*common.PinnedMessage, []string, error) {
+
+	var cursors []string
+	messageIdx := make(map[string]*common.PinnedMessage, 0)
+
+	for rows.Next() {
+		var (
+			message  common.Message
+			pinnedAt uint64
+			pinnedBy string
+			cursor   string
+		)
+		if err := db.tableUserMessagesScanAllFields(rows, &message, &pinnedAt, &pinnedBy, &cursor); err != nil {
+			return nil, nil, err
+		}
+		if msg, ok := messageIdx[message.ID]; !ok {
+			pinnedMessage := &common.PinnedMessage{
+				Message:  &message,
+				PinnedAt: pinnedAt,
+				PinnedBy: pinnedBy,
+			}
+			messageIdx[message.ID] = pinnedMessage
+			cursors = append(cursors, cursor)
+		} else if discordMessage := msg.Message.GetDiscordMessage(); discordMessage != nil {
+			msg.Message.Payload = getUpdatedChatMessagePayload(discordMessage, message.GetDiscordMessage())
+			messageIdx[message.ID] = msg
+		}
+	}
+
+	var messages []*common.PinnedMessage
+	for _, message := range messageIdx {
+		messages = append(messages, message)
+	}
+	sort.Sort(sort.Reverse(ByClockPinned{messages}))
+
+	return messages, cursors, nil
+}
```

Comment:
Very similar here: `getPinnedMessagesAndCursorsFromScanRows()` just operates on a different type (`common.PinnedMessage` instead of `common.Message`)

Any way to make this nicer in go?